include/rocksdb/utilities/cache_dump_load.h

```diff
@@ -0,0 +1,77 @@
+//  Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#pragma once
+
+#include <set>
+
+#include "rocksdb/cache.h"
+#include "rocksdb/env.h"
+#include "rocksdb/io_status.h"
+#include "rocksdb/secondary_cache.h"
+#include "rocksdb/table.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+class CacheDumpWriter {
+ public:
+  virtual ~CacheDumpWriter() = default;
+
+  virtual IOStatus Write(const Slice& data) = 0;
+  virtual IOStatus Close() = 0;
+  virtual uint64_t GetFileSize() = 0;
+};
+
+class CacheDumpReader {
+ public:
+  virtual ~CacheDumpReader() = default;
+  virtual IOStatus Read(size_t len, std::string* data) = 0;
+  virtual size_t GetOffset() const = 0;
+  virtual IOStatus Close() = 0;
+};
+
+// CacheDumpOptions is used to control DumpCache()
+struct CacheDumpOptions {
+  // The env pointer
+  Env* env;
+  // The prefix set, which filter out the cache keys that does not match
+  std::set<std::string> prefix_set;
+};
+
+class CacheDumper {
+ public:
+  virtual ~CacheDumper() = default;
+  virtual Status Prepare() = 0;
+  virtual IOStatus Run() = 0;
+};
+
+class CacheDumpedLoader {
+ public:
+  virtual ~CacheDumpedLoader() = default;
+  virtual Status Prepare() = 0;
+  virtual IOStatus Run() = 0;
+};
+
+Status NewDefaultCacheDumpWriter(Env* env, const EnvOptions& env_options,
+                                 const std::string& file_name,
+                                 std::unique_ptr<CacheDumpWriter>* writer);
+
+Status NewDefaultCacheDumpReader(Env* env, const EnvOptions& env_options,
+                                 const std::string& file_name,
+                                 std::unique_ptr<CacheDumpReader>* reader);
+
+Status NewDefaultCacheDumper(const CacheDumpOptions& dump_options,
+                             const std::shared_ptr<Cache>& cache,
+                             std::unique_ptr<CacheDumpWriter>&& writer,
+                             std::unique_ptr<CacheDumper>* cache_dumper);
+
+Status NewDefaultCacheDumpedLoader(
+    const CacheDumpOptions& dump_options,
+    const BlockBasedTableOptions& toptions,
+    const std::shared_ptr<SecondaryCache>& secondary_cache,
+    std::unique_ptr<CacheDumpReader>&& reader,
+    std::unique_ptr<CacheDumpedLoader>* cache_dump_loader);
+
+}  // namespace ROCKSDB_NAMESPACE
```

Comment:
This API doesn't seem very convenient for either the high level "simply dump/restore to/from file" use case nor very powerful/flexible for the low level "I'm going to custom transport my cache entry data" use case.

For high level "simply dump/restore to/from file," why can't there just be DumpCacheEntriesToFile and RestoreCacheEntriesFromFile functions? That seems much simpler without dealing with a bunch of objects and several statuses.

For low level "I'm going to custom transport my cache entry data," I think it gives the user much more power (without really complicating the interface) to provide individual serialized entries. When there is some kind of problem in sending a stream (e.g. repeated checksum mismatch due to out of sync), what can you do? You probably have to abandon the rest of the stream and either start over or forget about loading the rest of the entries. If there is a problem in sending individual or groups of entries, it is easy to retry each one or simply skip some and keep going. This would also make it easier to use parallel transports if desired. As long as the serialized entries form a prefix code (already contain their own length) then the user is free to concatenate them for processing on the other end. https://en.wikipedia.org/wiki/Prefix_code

It does complicate the "overall metadata" validation problem, because we don't want an elaborate header on each entry, but I expect it's fine to have a 32-bit version flag in each entry. (By the way, thanks for considering the versioning problem up front!)

Reply:
"why can't there just be DumpCacheEntriesToFile and RestoreCacheEntriesFromFile functions? That seems much simpler without dealing with a bunch of objects and several statuses."

Actually, combing Prepare() and Run() will achieve the DumpCacheEntries and RestoreCacheEntries as the one-time dumping and loading. 

Yes, in the next stage, we are considering to provide a "Step" loading API to user, in this way, the users is able to load blocks one-by-one. But that might need more flexibilities (e.g., check the block type, check the size, control the loading speed) and careful designs. 

In the current format design, I have considered the issues you mentioned and we are able to restart from any location (the APIs are not provided). Each block has its "sequence number". When we want to load a block with certain sequence number (e.g., some issues happens when we load block N), the reader can start from beginning, read metadata, decode it, get the offset of its data; jump to read the next metadata, decode it, get the offset off the following data......until we reach to block N. Since we are not a SST file or other "Tree" format file, we do not have the indexing structure stored. I think search linearly is already enough in this use case, since most likely, user just want to load them all.

Reply:
The "too many steps" problem looks more reasonable now that Prepare() and Run() are not separate.

Reply:
The Reader/Writer interfaces are clearly stream oriented. With that interface, you would have to depend on implementation details (subject to change?) to successfully make it a packet interface where packets can be re-ordered or dropped. I think it would be useful as a packet interface, something like this:

```
class CacheDumpWriter {
 public:
  virtual ~CacheDumpWriter() = default;

  // Called once before the calls to WritePacket
  virtual IOStatus WriteMetadata(const Slice& data) = 0;
  virtual IOStatus WritePacket(const Slice& data) = 0;
  virtual IOStatus Close() = 0;
};

class CacheDumpReader {
 public:
  virtual ~CacheDumpReader() = default;
  // Called once before the calls to ReadPacket
  virtual IOStatus ReadMetadata(std::string* data) = 0;
  // Sets data to empty string on EOF
  virtual IOStatus ReadPacket(std::string* data) = 0;
  // (Close not needed)
};
```

Reply:
Sure, in general, it will be clear to follow the metadata->data read pattern. One more question, in the future, if we want to support the case that we want to read out a certain block (due to failure or other cases), do we need a interface with `Read(size_t len, size_t offset, std::string* data)`? Also, `ReadMetadata(std::string* data)` and `ReadPacket(std::string* data)` requires we resize the string to a length before we call the API?

Reply:
> The "too many steps" problem looks more reasonable now that Prepare() and Run() are not separate.

I'm kind of confused by this comment. Can you explain a little more?

Reply:
Hi @pdillinger , can you provide more information?

Reply:
> if we want to support the case that we want to read out a certain block (due to failure or other cases), do we need a interface with Read(size_t len, size_t offset, std::string* data)?

If an error recovery piece is missing, I think it would be the ability for RocksDB to report to the CacheDumpReader that the last read did not pass sanity check. The CacheDumpReader would have the option of retrying, skipping, or aborting (on RocksDB's next call to ReadPacket / ReadMetadata).

> Also, ReadMetadata(std::string* data) and ReadPacket(std::string* data) requires we resize the string to a length before we call the API?

Why? We don't know how big the next data item will be before we have read it. The `std::string*` are intended purely as output parameters.

> The "too many steps" problem looks more reasonable now that Prepare() and Run() are not separate.

I previously claimed that it would be a lot of code just to simply "dump to file" or "restore from file." Part of the complexity was that it wasn't clear why Prepare() was separate from Run() or why or how you might handle errors from those steps differently. I'm now satisfied this has been fixed.